File: xju/io/test-FileWriter.cc
// Copyright (c) 2017 Trevor Taylor
//
// Permission to use, copy, modify, distribute and sell this software
// and its documentation for any purpose is hereby granted without fee,
// provided that the above copyright notice appear in all.
// Trevor Taylor makes no representations about the suitability of this
// software for any purpose.  It is provided "as is" without express or
// implied warranty.
//
#include <xju/io/FileWriter.hh>

#include <iostream>
#include <xju/assert.hh>
#include <xju/file/rm.hh>
#include <xju/file/read.hh>
#include <xju/file/touch.hh>

namespace xju
{
namespace io
{

void test1() {
  auto const fileName(xju::path::split("test.txt"));
  xju::file::touch(fileName,xju::file::Mode(0777));
  xju::file::rm(fileName);
  try{
    FileWriter x(fileName);
  }
  catch(xju::SyscallFailed const& e){
    xju::assert_equal(e._errno,ENOENT);
  }
  {
    FileWriter x(fileName,xju::file::Mode(0777),FileWriter::Exclusive());
    x.write("fred",4);
    x.write("jock",4);
  }
  xju::assert_equal(xju::file::read(fileName),
                    std::string("fredjock"));
  try{
    FileWriter x(fileName,xju::file::Mode(0777),FileWriter::Exclusive());
    xju::assert_never_reached();
  }
  catch(xju::SyscallFailed const& e){
    xju::assert_equal(e._errno,EEXIST);
  }
  {
    FileWriter x(fileName);
    x.write("sally",5);
    xju::assert_equal(x.seekBy(-4),9);
    x.write("i",1);
  }
  xju::assert_equal(xju::file::read(fileName),
                    std::string("fredjocksilly"));
  {
    xju::file::rm(fileName);
    FileWriter x(fileName,xju::file::Mode(0777));
    x.write("fred",4);
    x.write("jock",4);
  }
  xju::assert_equal(xju::file::read(fileName),
                    std::string("fredjock"));
  {
    FileWriter x(fileName);
    x.write("sally",5);
    xju::assert_equal(&x.seekTo(1),&x);
    x.write("e",1);
  }
  xju::assert_equal(xju::file::read(fileName),
                    std::string("feedjocksally"));
  {
    FileWriter x(fileName);
    x.truncate(4);
    x.write("sally",5);

    try{
      x.seekTo(-1);
      xju::assert_never_reached();
    }
    catch(xju::SyscallFailed const& e){
      xju::assert_equal(e._errno,EINVAL);
    }
    
  }
  xju::assert_equal(xju::file::read(fileName),
                    std::string("feedsally"));
}

}
}

using namespace xju::io;

int main(int argc, char* argv[])
{
  unsigned int n(0);
  test1(), ++n;
  std::cout << "PASS - " << n << " steps" << std::endl;
  return 0;
}

File: xju/cmc/io/test-FileWriter.py
#!/usr/bin/env python3

# Copyright (c) 2022 Trevor Taylor
# coding: utf-8
# 
# Permission to use, copy, modify, and/or distribute this software for
# any purpose with or without fee is hereby granted, provided that all
# copyright notices and this permission notice appear in all copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#


from xju.cmc.io import FileWriter,FileReader,FileMode
from pathlib import Path
from xju.assert_ import Assert
from xju.xn import readable_repr
from xju.misc import ByteCount

try:
    with FileWriter(Path("xxx.txt")) as f:
        pass
except Exception as e:
    Assert(readable_repr(e))=="Failed to open file writer for xxx.txt with with create mode None, must not exist False, close-on-exec True because\n[Errno 2] No such file or directory: 'xxx.txt'."
else:
    assert False, f'should not be here with {f}'
    pass

with FileWriter(Path('xxx.txt'),mode=FileMode(0o666)) as f:
    f.output.write(b'fred')
    Assert(f.size())==ByteCount(4)
    f.output.write(b'ward')
    pass

try:
    with FileWriter(Path("xxx.txt"), mode=FileMode(0o666), must_not_exist=True) as f:
        pass
except FileExistsError as e:
    Assert(readable_repr(e))=="Failed to open file writer for xxx.txt with with create mode 0o666, must not exist True, close-on-exec True because\n[Errno 17] File exists: 'xxx.txt'."
else:
    assert False, f'should not be here with {f}'
    pass

with FileReader(Path('xxx.txt')) as f2:
    Assert(f2.input.readall())==b'fredward'
    pass

with FileWriter(Path("xxx.txt")) as f:
    f.seek_by(4)
    f.output.write(b'lea')
    f.seek_to(1)
    f.output.write(b'i')
    pass

assert not hasattr(f,'output')

with FileReader(Path('xxx.txt')) as f2:
    Assert(f2.input.readall())==b'fiedlead'
    pass
